### DISCUSSION TITLE
Unable to generate project using mwc.pl

### DISCUSSION THREAD

**Discussion by axuchao458** [OpenDDS/OpenDDS#4870] [Q&A] [answered]
Hello, Professor：
I use mwc.pl type gnuace to generate the project, but there is an error message stating that there is no option for gnuace
![9f16b117d6bdc4628d9e7890ad87399](https://github.com/user-attachments/assets/ebac5f29-6e46-49ea-868d-9559d72fac0f)
If I change the command to mwc.pl type make, there will be an error saying 'unable locate parent: dcps_java' again
![image](https://github.com/user-attachments/assets/acf0f605-362d-4859-b6bd-2deaf432c2c8)
'dcps_java' is in my MPC file
![image](https://github.com/user-attachments/assets/04e91934-4cf8-4672-8704-2e2d187ba2cc)

My system is Ubuntu 22.04, OpenDDS is 3.30, Java is 21, and the reference is https://opendds.readthedocs.io/en/latest-release/devguide/quickstart/linux.html

**Comment by mitza-oci** [comment]
Run the command ". setenv.sh" (with the current directory set to the root of the OpenDDS source tree) before running "mwc.pl".  This will set the environment variables needed for MPC.
If that doesn't work, run "tools/scripts/show_build_config.pl" and paste the plain text output here in a comment

**Reply by axuchao458** [reply]
Thank you very much for your help
I executed 'source setenv. sh' in the root directory of OpenDDS and then generated the jar.
But why is there no class file in the jar I generated?
![image](https://github.com/user-attachments/assets/b71d1f1a-4a0b-4e7d-9753-d55df0e23208)

**Reply by mitza-oci** [reply]
Please send the plain text log from building this library.

**Reply by axuchao458** [reply]
Hello, Professor
I have attached a text attachment
mwc.txt is the output of mwc
[mwc.txt](https://github.com/user-attachments/files/18234996/mwc.txt)
make.txt  is the output of make
[make.txt](https://github.com/user-attachments/files/18234998/make.txt)
There is still no class file in the jar I generated

**Reply by jrw972** [reply]
Were you able to resolve your issue?

**Reply by axuchao458** [reply]
not yet

**Reply by jrw972** [reply]
Please run "tools/scripts/show_build_config.pl" and paste the plain text output here in a comment

**Reply by axuchao458** [reply]
Thank you very much for your help
[build_config.txt](https://github.com/user-attachments/files/18483596/build_config.txt)

**Reply by jrw972** [reply]
Assuming you have a Makefile, which it appears you do, please try the following:
1. Start a new terminal.
2. `cd ~/Documents/opendds-liunx-test-202412181610`
3. `. /home/vboxuser/snap/OpenDDS-3.30/setenv.sh`
4. `make realclean`
6. `opendds_mwc.pl`
7. `make`

Please capture all of the output into a single log file for analysis.

**Reply by axuchao458** [reply]
Sorry, I am not very proficient in Liunx. I may encounter errors when following the steps
![d020c0da6664fcb9d094bad3e708f0e](https://github.com/user-attachments/assets/475e4ab4-5988-4f8f-80de-9576012638e1)

**Reply by jrw972** [reply]
Sorry.  I missed a step.  I'll correct the step in my previous comment.

**Reply by axuchao458** [reply]
Thank you very much for your help. It can now be executed
1.. /home/vboxuser/snap/OpenDDS-3.30/setenv.sh
    no-output
2.make realclean
    
[realclean.txt](https://github.com/user-attachments/files/18529382/realclean.txt)
3.opendds_mwc.pl
    
[opendds_mwc.txt](https://github.com/user-attachments/files/18529386/opendds_mwc.txt)
4.make
    
[make.txt](https://github.com/user-attachments/files/18529391/make.txt)

**Reply by jrw972** [reply]
Post your IDL file.  I have a feeling you didn't define a topic or something like that.

**Reply by axuchao458** [reply]
Okay, thank you very much
[YCDef.txt](https://github.com/user-attachments/files/18665238/YCDef.txt)

**Reply by jrw972** [reply]
There is an issue here but I was able to work around it.
Essentially, run opendds_mwc.pl (or mwc.pl), then make.
Then run opendds_mwc.pl (or mwc.pl), then make again.
For some reason, it's not picking up the generated .java files.
Doing this, I was able to compile your IDL with your .mpc file.

**Reply by axuchao458** [reply]
Sorry, I don't quite understand. Do I need to compile twice?

**Reply by jrw972** [reply, marked as answer]
Yes. You should try running `opendds_mwc.pl`, then `make`, then `opendds_mwc.pl`, and then `make` again to see if that fixes your problem.

**Reply by axuchao458** [reply]
Thanks a lot, with two compilations, I can get the jar

**Reply by mitza-oci** [reply]
@axuchao458 see https://github.com/OpenDDS/OpenDDS/issues/4886#issuecomment-2679734117 for more information on this
